File: tests/cpp/test_kernel_column.cpp
// Conformance fixture (C++ side): a COLUMN kernel — the mirror of
// tests/f90/test_kernel_column's `column_sum`, distilled from TIM's
// zonal_BT_mass_flux. The kernel is the body of a `ParallelFor` lambda over
// the horizontal cell (i, j): a per-column accumulator initialized to zero, a
// `for k` loop calling a banked point primitive with two `Real&` outputs
// received in uninitialized locals, a `+=` accumulation, and a per-column
// store `qbt(i,j,0)`. Reads `u(i,j,k)` are per-k arrays, `h(i+1,j,k)` a
// stencil, `dy(i,j,0)` a per-column scalar, `cs.vol_cfl` a member read on a
// const-reference struct parameter (a Bool input), and `specified_bc` a
// captured flag whose guarded block is pruned under the manifest hypothesis
// `specified_bc = false`. The prelude mirrors the AMReX shapes involved —
// `Array4::operator()`, `Box::smallEnd/bigEnd`, `ParallelFor(Box, lambda)` —
// so the fixture is self-contained (no includes; clang++ alone suffices).

using Real = double;
constexpr Real operator""_rt(long double v) { return static_cast<Real>(v); }
namespace amrex {
struct Box {
    int lo[3]; int hi[3];
    int smallEnd(int d) const noexcept { return lo[d]; }
    int bigEnd(int d) const noexcept { return hi[d]; }
};
template <class T> struct Array4 {
    T* p; long jstride; long kstride;
    T& operator()(int i, int j, int k) const noexcept { return p[i + j*jstride + k*kstride]; }
};
template <class F> void ParallelFor(const Box& bx, F&& f) noexcept {
    for (int k = bx.lo[2]; k <= bx.hi[2]; ++k)
        for (int j = bx.lo[1]; j <= bx.hi[1]; ++j)
            for (int i = bx.lo[0]; i <= bx.hi[0]; ++i) f(i, j, k);
}
}
using amrex::Box;
using amrex::Array4;
using amrex::ParallelFor;

struct cs_t { bool vol_cfl; };

// The banked point primitive the column kernel calls (two Real& outputs).
void flux_pt_point(Real const u, Real const h, Real const h_p1, Real& q, Real& dq,
                   Real const dy, Real const iarea, Real const iarea_p1, Real const dt,
                   bool const vol_cfl) noexcept
{
    Real cfl;
    if (vol_cfl) { cfl = u * dt * iarea; } else { cfl = u * dt * iarea_p1; }
    q = dy * u * (h + cfl * (h_p1 - h));
    dq = dy * h;
}

// Supported: the column kernel (ParallelFor #1 of this function).
void column_sum(const Box& bx, Array4<const Real> const& u, Array4<const Real> const& h,
                Array4<Real> const& qbt, Real dt, Array4<const Real> const& dy,
                Array4<const Real> const& iarea, const cs_t& cs, bool specified_bc)
{
    const int kmin = bx.smallEnd(2);
    const int kmax = bx.bigEnd(2);
    ParallelFor(bx, [=] (int i, int j, int) noexcept
    {
        Real qbt_val = 0.0_rt;
        for (int k = kmin; k <= kmax; ++k) {
            Real q_val, dq_val;
            flux_pt_point(u(i,j,k), h(i,j,k), h(i+1,j,k), q_val, dq_val,
                          dy(i,j,0), iarea(i,j,0), iarea(i+1,j,0), dt, cs.vol_cfl);
            if (specified_bc) { q_val = q_val / 2; }   // pruned: never modeled (an int literal would refuse)
            qbt_val += q_val;
        }
        qbt(i,j,0) = qbt_val;
    });
}

// REFUSAL: a per-k write that depends on the fold state — a scan.
void refuse_scan(const Box& bx, Array4<const Real> const& u, Array4<Real> const& w)
{
    const int kmin = bx.smallEnd(2);
    const int kmax = bx.bigEnd(2);
    ParallelFor(bx, [=] (int i, int j, int) noexcept
    {
        Real acc = 0.0_rt;
        for (int k = kmin; k <= kmax; ++k) {
            acc += u(i,j,k);
            w(i,j,k) = acc;
        }
    });
}

// REFUSAL: a call statement to a function that is not a banked primitive.
void helper(Real& acc, Real const x) noexcept { acc = acc + x; }
void refuse_unbanked_call(const Box& bx, Array4<const Real> const& u, Array4<Real> const& s)
{
    const int kmin = bx.smallEnd(2);
    const int kmax = bx.bigEnd(2);
    ParallelFor(bx, [=] (int i, int j, int) noexcept
    {
        Real acc = 0.0_rt;
        for (int k = kmin; k <= kmax; ++k) { helper(acc, u(i,j,k)); }
        s(i,j,0) = acc;
    });
}

File: tests/f90/test_kernel_column.f90
! A COLUMN kernel, distilled from MOM_continuity_PPM's zonal_BT_mass_flux:
! pointized over the column indices (j, i), the body is a sequence of
! k-passes — a `do concurrent (k,j,i)` MAP calling a banked point primitive
! (two intent(out) actuals land in per-k arrays), then a plain `do k;j;i`
! FOLD accumulating a per-column total. Around them, the shapes the real
! routine carries: a timer call (declared ignorable in the manifest), integer
! locals that only feed loop bounds, a whole-array assignment initializing the
! per-column output, and an OBC-style block guarded by a flag the manifest
! assumes false (its body would refuse if modeled — it reads q(i-1,j,k)).
! The siblings pin the boundary of the fold model.
module test_kernel_column
  implicit none

  type :: grid_t
    real(8), allocatable :: dy(:,:), iarea(:,:)
  end type grid_t

  type :: cs_t
    logical :: vol_cfl
  end type cs_t

contains

  subroutine timer_start()
  end subroutine timer_start

  subroutine timer_end()
  end subroutine timer_end

  ! The banked point primitive (a per-point elemental subroutine).
  elemental subroutine flux_pt(u, h, h_p1, q, dq, dy, iarea, iarea_p1, dt, vol_cfl)
    real(8), intent(in) :: u, h, h_p1, dy, iarea, iarea_p1, dt
    logical, intent(in) :: vol_cfl
    real(8), intent(out) :: q, dq
    real(8) :: cfl
    if (vol_cfl) then ; cfl = u * dt * iarea ; else ; cfl = u * dt * iarea_p1 ; endif
    q = dy * u * (h + cfl * (h_p1 - h))
    dq = dy * h
  end subroutine flux_pt

  ! Supported: the column kernel (columns = j, i; assume specified_bc = false;
  ! ignore_calls = timer_start, timer_end).
  subroutine column_sum(u, h, qbt, dt, g, cs, n, m, nz, specified_bc)
    integer, intent(in) :: n, m, nz
    real(8), intent(in) :: u(0:n, m, nz), h(0:n+1, m, nz)
    real(8), intent(out) :: qbt(0:n, m)
    real(8), intent(in) :: dt
    type(grid_t), intent(in) :: g
    type(cs_t), intent(in) :: cs
    logical, intent(in) :: specified_bc
    real(8) :: q(0:n, m, nz), dq(0:n, m, nz)
    integer :: i, j, k, ie
    call timer_start()
    ie = n
    qbt(:,:) = 0.0
    do concurrent (k=1:nz, j=1:m, i=0:ie)
      call flux_pt(u(i,j,k), h(i,j,k), h(i+1,j,k), q(i,j,k), dq(i,j,k), &
                   g%dy(i,j), g%iarea(i,j), g%iarea(i+1,j), dt, cs%vol_cfl)
      if (specified_bc) q(i,j,k) = q(i-1,j,k)
    end do
    do k = 1, nz ; do j = 1, m ; do i = 0, ie
      qbt(i,j) = qbt(i,j) + q(i,j,k)
    end do ; end do ; end do
    call timer_end()
  end subroutine column_sum

  ! REFUSAL: a per-k write that depends on the fold state — a scan.
  subroutine scan(u, w, n, m, nz)
    integer, intent(in) :: n, m, nz
    real(8), intent(in) :: u(n, m, nz)
    real(8), intent(out) :: w(n, m, nz)
    real(8) :: acc(n, m)
    integer :: i, j, k
    acc(:,:) = 0.0
    do k = 1, nz ; do j = 1, m ; do i = 1, n
      acc(i,j) = acc(i,j) + u(i,j,k)
      w(i,j,k) = acc(i,j)
    end do ; end do ; end do
  end subroutine scan

  ! REFUSAL: a read of a per-k array at an offset in k — a k-recurrence.
  subroutine k_recurrence(u, w, n, m, nz)
    integer, intent(in) :: n, m, nz
    real(8), intent(in) :: u(n, m, nz)
    real(8), intent(inout) :: w(n, m, 0:nz)
    integer :: i, j, k
    do k = 1, nz ; do j = 1, m ; do i = 1, n
      w(i,j,k) = w(i,j,k-1) + u(i,j,k)
    end do ; end do ; end do
  end subroutine k_recurrence

  ! REFUSAL: a call to a procedure that is neither banked nor declared ignorable.
  subroutine unbanked_call(u, s, n, m, nz)
    integer, intent(in) :: n, m, nz
    real(8), intent(in) :: u(n, m, nz)
    real(8), intent(out) :: s(n, m)
    integer :: i, j, k
    s(:,:) = 0.0
    do k = 1, nz ; do j = 1, m ; do i = 1, n
      call timer_start()
      s(i,j) = s(i,j) + u(i,j,k)
    end do ; end do ; end do
  end subroutine unbanked_call

end module test_kernel_column
